# Part-1

Q1) Import required libraries and read the dataset.

In [98]:
import numpy as np                                      # for mathematical calculations
import pandas as pd                                     # for import and exploratory analysis of data
import matplotlib.pyplot as plt                         # for data visualization                     
import seaborn as sns                                   # for data visualization  
%matplotlib inline
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
appsdata = pd.read_csv("Apps_data.csv")                 #importing data
appsdata

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

Q2) Check the first few samples, shape, info of the data and try to familiarize yourself with different features.

In [3]:
appsdata.head()             #checking the first 5 rows in data

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [4]:
appsdata.tail()           #checking the last 5 rows in data

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

In [5]:
appsdata.shape            #checking shape of data

(10841, 13)

In [6]:
appsdata.info()          #info of data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


Q3) Check summary statistics of the dataset. List out the columns that need to be worked upon for model building.

In [7]:
appsdata.describe(include = 'all')

App Category       Rating Reviews                Size    Installs  \
count    10841    10841  9367.000000   10841               10841       10841   
unique    9660       34          NaN    6002                 462          22   
top     ROBLOX   FAMILY          NaN       0  Varies with device  1,000,000+   
freq         9     1972          NaN     596                1695        1579   
mean       NaN      NaN     4.193338     NaN                 NaN         NaN   
std        NaN      NaN     0.537431     NaN                 NaN         NaN   
min        NaN      NaN     1.000000     NaN                 NaN         NaN   
25%        NaN      NaN     4.000000     NaN                 NaN         NaN   
50%        NaN      NaN     4.300000     NaN                 NaN         NaN   
75%        NaN      NaN     4.500000     NaN                 NaN         NaN   
max        NaN      NaN    19.000000     NaN                 NaN         NaN   

         Type  Price Content Rating Genres    Last Updated  \
count   10840  10841          10840  10841           10841   
unique      3     93              6    120            1378   
top      Free      0       Everyone  Tools  August 3, 2018   
freq    10039  10040           8714    842             326   
mean      NaN    NaN            NaN    NaN             NaN   
std       NaN    NaN            NaN    NaN             NaN   
min       NaN    NaN            NaN    NaN             NaN   
25%       NaN    NaN            NaN    NaN             NaN   
50%       NaN    NaN            NaN    NaN             NaN   
75%       NaN    NaN            NaN    NaN             NaN   
max       NaN    NaN            NaN    NaN             NaN   

               Current Ver Android Ver  
count                10833       10838  
unique                2832          33  
top     Varies with device  4.1 and up  
freq                  1459        2451  
mean                   NaN         NaN  
std                    NaN         NaN  
min                    NaN         NaN  
25%                    NaN         NaN  
50%                    NaN         NaN  
75%                    NaN         NaN  
max                    NaN         NaN

The columns to include in the model building would be:

Category

Rating

Size

Type

Price

Content Rating

Genres

Current Ver

Android Ver

Q4) Check if there are any duplicate records in the dataset? if any drop them.

In [8]:
appsdata.duplicated().sum()        #finding the number of duplicate values

483

There are 483 duplicate values in the data

In [9]:
appsdata.drop_duplicates(inplace = True)         # dropping the duplicate values

In [10]:
appsdata

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10358 rows x 13 columns]

After dropping duplicates, the number of rows are 10358

Q5) Check the unique categories of the column 'Category', Is there any invalid category? If yes, drop them.

In [11]:
appsdata['Category'].unique()    # To find the unique values in the 'Category' column

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

In [12]:
invalid = appsdata[(appsdata['Category']=='1.9')]
invalid

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [13]:
appsdata.drop([10472], inplace = True)

In [14]:
appsdata

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10357 rows x 13 columns]

Q6) Check if there are missing values present in the column Rating, If any? drop them and and create a new column as 'Rating_category' by converting ratings to high and low categories(>3.5 is high rest low)

In [15]:
appsdata[(appsdata['Rating'].isnull())]

App             Category  Rating  \
23                  Mcqueen Coloring pages       ART_AND_DESIGN     NaN   
113              Wrinkles and rejuvenation               BEAUTY     NaN   
123                 Manicure - nail design               BEAUTY     NaN   
126           Skin Care and Natural Beauty               BEAUTY     NaN   
129    Secrets of beauty, youth and health               BEAUTY     NaN   
...                                    ...                  ...     ...   
10824                            Cardio-FR              MEDICAL     NaN   
10825                   Naruto & Boruto FR               SOCIAL     NaN   
10831             payermonstationnement.fr  MAPS_AND_NAVIGATION     NaN   
10835                             FR Forms             BUSINESS     NaN   
10838               Parkinson Exercices FR              MEDICAL     NaN   

      Reviews  Size  Installs  Type Price Content Rating  \
23         61  7.0M  100,000+  Free     0       Everyone   
113       182  5.7M  100,000+  Free     0   Everyone 10+   
123       119  3.7M   50,000+  Free     0       Everyone   
126       654  7.4M  100,000+  Free     0           Teen   
129        77  2.9M   10,000+  Free     0     Mature 17+   
...       ...   ...       ...   ...   ...            ...   
10824      67   82M   10,000+  Free     0       Everyone   
10825       7  7.7M      100+  Free     0           Teen   
10831      38  9.8M    5,000+  Free     0       Everyone   
10835       0  9.6M       10+  Free     0       Everyone   
10838       3  9.5M    1,000+  Free     0       Everyone   

                                Genres        Last Updated Current Ver  \
23     Art & Design;Action & Adventure       March 7, 2018       1.0.0   
113                             Beauty  September 20, 2017         8.0   
123                             Beauty       July 23, 2018         1.3   
126                             Beauty       July 17, 2018        1.15   
129                             Beauty      August 8, 2017         2.0   
...                                ...                 ...         ...   
10824                          Medical       July 31, 2018       2.2.2   
10825                           Social    February 2, 2018         1.0   
10831                Maps & Navigation       June 13, 2018   2.0.148.0   
10835                         Business  September 29, 2016       1.1.5   
10838                          Medical    January 20, 2017         1.0   

      Android Ver  
23     4.1 and up  
113    3.0 and up  
123    4.1 and up  
126    4.1 and up  
129    2.3 and up  
...           ...  
10824  4.4 and up  
10825  4.0 and up  
10831  4.0 and up  
10835  4.0 and up  
10838  2.2 and up  

[1465 rows x 13 columns]

In [16]:
appsdata.dropna(subset = 'Rating',inplace=True)

In [17]:
appsdata

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10834     4.0       7                2.6M         500+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10834       Everyone                  Education     June 18, 2017   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10834               1.0.0          4.1 and up  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[8892 rows x 13 columns]

In [18]:
appsdata['Rating_category'] = appsdata['Rating'].apply(lambda x: 'High' if x>3.5 else 'Low')

In [19]:
appsdata

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10834     4.0       7                2.6M         500+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10834       Everyone                  Education     June 18, 2017   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver Rating_category  
0                   1.0.0        4.0.3 and up            High  
1                   2.0.0        4.0.3 and up            High  
2                   1.2.4        4.0.3 and up            High  
3      Varies with device          4.2 and up            High  
4                     1.1          4.4 and up            High  
...                   ...                 ...             ...  
10834               1.0.0          4.1 and up            High  
10836                1.48          4.1 and up            High  
10837                 1.0          4.1 and up            High  
10839  Varies with device  Varies with device            High  
10840  Varies with device  Varies with device            High  

[8892 rows x 14 columns]

Q7) Check the distribution of the newly created column 'Rating_category' and comment on the distribution.

In [20]:
appsdata['Rating_category'].value_counts()

High    8012
Low      880
Name: Rating_category, dtype: int64

The above counts show that 8012(90%) apps have a high-rating(> 3.5) while 880(10%) apps have a low rating(< =  3.5). 

Q8) Convert the column "Reviews'' to numeric data type and check the presence of outliers in the column and handle the outliers using a transformation approach.

In [23]:
appsdata.dtypes

App                 object
Category            object
Rating             float64
Reviews             object
Size                object
Installs            object
Type                object
Price               object
Content Rating      object
Genres              object
Last Updated        object
Current Ver         object
Android Ver         object
Rating_category     object
dtype: object

In [24]:
appsdata['Reviews'].dtype

dtype('O')

In [35]:
appsdata['Reviews'] = appsdata['Reviews'].astype(int)

In [36]:
appsdata['Reviews'].dtype

dtype('int32')

In [46]:
Q1 = appsdata['Reviews'].quantile(0.25)

In [47]:
Q3 = appsdata['Reviews'].quantile(0.75)

In [48]:
Q2 = appsdata['Reviews'].median()

In [50]:
x = 1.5*(Q3-Q1)
x

106654.125

In [52]:
appsdata[(appsdata['Reviews'] < (Q1-x)) | (appsdata['Reviews'] > (Q3+x))]

App           Category  \
3                              Sketch - Draw & Paint     ART_AND_DESIGN   
18                     FlipaClip - Cartoon animation     ART_AND_DESIGN   
19                                      ibis Paint X     ART_AND_DESIGN   
42                        Textgram - write on photos     ART_AND_DESIGN   
72     Android Auto - Maps, Media, Messaging & Voice  AUTO_AND_VEHICLES   
...                                              ...                ...   
10740                                     PhotoFunia        PHOTOGRAPHY   
10781                           Modern Strike Online               GAME   
10784                                     Big Hunter               GAME   
10809          Castle Clash: RPG War and Strategy FR             FAMILY   
10840  iHoroscope - 2018 Daily Horoscope & Astrology          LIFESTYLE   

       Rating  Reviews                Size     Installs  Type Price  \
3         4.5   215644                 25M  50,000,000+  Free     0   
18        4.3   194216                 39M   5,000,000+  Free     0   
19        4.6   224399                 31M  10,000,000+  Free     0   
42        4.4   295221  Varies with device  10,000,000+  Free     0   
72        4.2   271920                 16M  10,000,000+  Free     0   
...       ...      ...                 ...          ...   ...   ...   
10740     4.3   316378                4.4M  10,000,000+  Free     0   
10781     4.3   834117                 44M  10,000,000+  Free     0   
10784     4.3   245455                 84M  10,000,000+  Free     0   
10809     4.7   376223                 24M   1,000,000+  Free     0   
10840     4.5   398307                 19M  10,000,000+  Free     0   

      Content Rating           Genres    Last Updated         Current Ver  \
3               Teen     Art & Design    June 8, 2018  Varies with device   
18          Everyone     Art & Design  August 3, 2018               2.2.5   
19          Everyone     Art & Design   July 30, 2018               5.5.4   
42          Everyone     Art & Design   July 30, 2018  Varies with device   
72              Teen  Auto & Vehicles   July 11, 2018  Varies with device   
...              ...              ...             ...                 ...   
10740       Everyone      Photography    June 3, 2017             4.0.7.0   
10781           Teen           Action   July 30, 2018              1.25.4   
10784   Everyone 10+           Action    May 31, 2018               2.8.6   
10809       Everyone         Strategy   July 18, 2018               1.4.2   
10840       Everyone        Lifestyle   July 25, 2018  Varies with device   

              Android Ver Rating_category  
3              4.2 and up            High  
18           4.0.3 and up            High  
19             4.1 and up            High  
42     Varies with device            High  
72             5.0 and up            High  
...                   ...             ...  
10740          2.3 and up            High  
10781          4.1 and up            High  
10784          4.0 and up            High  
10809          4.1 and up            High  
10840  Varies with device            High  

[1555 rows x 14 columns]

The above analysis shows that there are 1555 outliers present in the data

Using log transformation to handle outliers

In [58]:
appsdata['Reviews'] = np.log(appsdata['Reviews'])

In [59]:
q1 = appsdata['Reviews'].quantile(0.25)
q3 = appsdata['Reviews'].quantile(0.75)
y = 1.5*(q3-q1)
appsdata[(appsdata['Reviews'] < (q1-y)) | (appsdata['Reviews'] > (q3+y))]

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver, Rating_category]
Index: []

We can see that there are no more outliars present in the data

Q9) The column 'Size' contains alphanumeric values, treat the non numeric data and convert the column into suitable data type.

In [74]:
data = appsdata.copy()
data

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating    Reviews                Size     Installs  Type Price  \
0         4.1   5.068904                 19M      10,000+  Free     0   
1         3.9   6.874198                 14M     500,000+  Free     0   
2         4.7  11.379508                8.7M   5,000,000+  Free     0   
3         4.5  12.281384                 25M  50,000,000+  Free     0   
4         4.3   6.874198                2.8M     100,000+  Free     0   
...       ...        ...                 ...          ...   ...   ...   
10834     4.0   1.945910                2.6M         500+  Free     0   
10836     4.5   3.637586                 53M       5,000+  Free     0   
10837     5.0   1.386294                3.6M         100+  Free     0   
10839     4.5   4.736198  Varies with device       1,000+  Free     0   
10840     4.5  12.894978                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10834       Everyone                  Education     June 18, 2017   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver Rating_category  
0                   1.0.0        4.0.3 and up            High  
1                   2.0.0        4.0.3 and up            High  
2                   1.2.4        4.0.3 and up            High  
3      Varies with device          4.2 and up            High  
4                     1.1          4.4 and up            High  
...                   ...                 ...             ...  
10834               1.0.0          4.1 and up            High  
10836                1.48          4.1 and up            High  
10837                 1.0          4.1 and up            High  
10839  Varies with device  Varies with device            High  
10840  Varies with device  Varies with device            High  

[8892 rows x 14 columns]

In [75]:
len(data[(data['Size']=='Varies with device')])

1468

In [76]:
#1 dropping rows with size as 'Varies with device'

data.drop(data[data['Size']=='Varies with device'].index, inplace = True)

In [77]:
len(data[(data['Size']=='Varies with device')])

0

In [78]:
#2 converting M to 1000000 and k to 1000
num = {'M':1000000, 'k':1000}

data['Size'] = (data['Size'].str[0:-1].astype(float)*data['Size'].str[-1].replace(num)).astype(float)

In [79]:
data

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10833                                        Chemin (fr)  BOOKS_AND_REFERENCE   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating    Reviews        Size     Installs  Type Price Content Rating  \
0         4.1   5.068904  19000000.0      10,000+  Free     0       Everyone   
1         3.9   6.874198  14000000.0     500,000+  Free     0       Everyone   
2         4.7  11.379508   8700000.0   5,000,000+  Free     0       Everyone   
3         4.5  12.281384  25000000.0  50,000,000+  Free     0           Teen   
4         4.3   6.874198   2800000.0     100,000+  Free     0       Everyone   
...       ...        ...         ...          ...   ...   ...            ...   
10833     4.8   3.784190    619000.0       1,000+  Free     0       Everyone   
10834     4.0   1.945910   2600000.0         500+  Free     0       Everyone   
10836     4.5   3.637586  53000000.0       5,000+  Free     0       Everyone   
10837     5.0   1.386294   3600000.0         100+  Free     0       Everyone   
10840     4.5  12.894978  19000000.0  10,000,000+  Free     0       Everyone   

                          Genres      Last Updated         Current Ver  \
0                   Art & Design   January 7, 2018               1.0.0   
1      Art & Design;Pretend Play  January 15, 2018               2.0.0   
2                   Art & Design    August 1, 2018               1.2.4   
3                   Art & Design      June 8, 2018  Varies with device   
4        Art & Design;Creativity     June 20, 2018                 1.1   
...                          ...               ...                 ...   
10833          Books & Reference    March 23, 2014                 0.8   
10834                  Education     June 18, 2017               1.0.0   
10836                  Education     July 25, 2017                1.48   
10837                  Education      July 6, 2018                 1.0   
10840                  Lifestyle     July 25, 2018  Varies with device   

              Android Ver Rating_category  
0            4.0.3 and up            High  
1            4.0.3 and up            High  
2            4.0.3 and up            High  
3              4.2 and up            High  
4              4.4 and up            High  
...                   ...             ...  
10833          2.2 and up            High  
10834          4.1 and up            High  
10836          4.1 and up            High  
10837          4.1 and up            High  
10840  Varies with device            High  

[7424 rows x 14 columns]

Q10) Check the column 'Installs', treat the unwanted characters and convert the column into a suitable data type.

In [81]:
len(data['Installs'].unique())

19

In [82]:
data['Installs'].unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000+', '500,000,000+', '100+', '500+', '10+', '1,000,000,000+',
       '5+', '50+', '1+'], dtype=object)

In [84]:
data['Installs'] = data['Installs'].map(lambda x: x.rstrip('+'))
data['Installs'] = data['Installs'].map(lambda x: ''.join(x.split(',')))
data['Installs']

0           10000
1          500000
2         5000000
3        50000000
4          100000
           ...   
10833        1000
10834         500
10836        5000
10837         100
10840    10000000
Name: Installs, Length: 7424, dtype: object

In [85]:
data['Installs'] = data['Installs'].apply(lambda x : x.replace(',','').replace('+','')).astype(int)

Q11) Check the column 'Price' , remove the unwanted characters and convert the column into a suitable data type.

In [89]:
data['Price'] = data['Price'].apply(lambda x: x.lstrip('$').rstrip(" ")).astype(float)

In [90]:
data

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10833                                        Chemin (fr)  BOOKS_AND_REFERENCE   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating    Reviews        Size  Installs  Type  Price Content Rating  \
0         4.1   5.068904  19000000.0     10000  Free    0.0       Everyone   
1         3.9   6.874198  14000000.0    500000  Free    0.0       Everyone   
2         4.7  11.379508   8700000.0   5000000  Free    0.0       Everyone   
3         4.5  12.281384  25000000.0  50000000  Free    0.0           Teen   
4         4.3   6.874198   2800000.0    100000  Free    0.0       Everyone   
...       ...        ...         ...       ...   ...    ...            ...   
10833     4.8   3.784190    619000.0      1000  Free    0.0       Everyone   
10834     4.0   1.945910   2600000.0       500  Free    0.0       Everyone   
10836     4.5   3.637586  53000000.0      5000  Free    0.0       Everyone   
10837     5.0   1.386294   3600000.0       100  Free    0.0       Everyone   
10840     4.5  12.894978  19000000.0  10000000  Free    0.0       Everyone   

                          Genres      Last Updated         Current Ver  \
0                   Art & Design   January 7, 2018               1.0.0   
1      Art & Design;Pretend Play  January 15, 2018               2.0.0   
2                   Art & Design    August 1, 2018               1.2.4   
3                   Art & Design      June 8, 2018  Varies with device   
4        Art & Design;Creativity     June 20, 2018                 1.1   
...                          ...               ...                 ...   
10833          Books & Reference    March 23, 2014                 0.8   
10834                  Education     June 18, 2017               1.0.0   
10836                  Education     July 25, 2017                1.48   
10837                  Education      July 6, 2018                 1.0   
10840                  Lifestyle     July 25, 2018  Varies with device   

              Android Ver Rating_category  
0            4.0.3 and up            High  
1            4.0.3 and up            High  
2            4.0.3 and up            High  
3              4.2 and up            High  
4              4.4 and up            High  
...                   ...             ...  
10833          2.2 and up            High  
10834          4.1 and up            High  
10836          4.1 and up            High  
10837          4.1 and up            High  
10840  Varies with device            High  

[7424 rows x 14 columns]

In [92]:
data['Price'].dtype

dtype('float64')

Q12) Drop the columns which you think redundant for the analysis.(suggestion: drop column 'rating', since we created a new feature from it (i.e. rating_category) and the columns 'App', 'Rating' ,'Genres','Last Updated', 'Current Ver','Android Ver' columns since which are redundant for our analysis) 

In [93]:
data1 = data.copy()

In [96]:
data1.drop(['App', 'Rating' ,'Genres','Last Updated', 'Current Ver','Android Ver'], axis=1, inplace=True)
data1

Category    Reviews        Size  Installs  Type  Price  \
0           ART_AND_DESIGN   5.068904  19000000.0     10000  Free    0.0   
1           ART_AND_DESIGN   6.874198  14000000.0    500000  Free    0.0   
2           ART_AND_DESIGN  11.379508   8700000.0   5000000  Free    0.0   
3           ART_AND_DESIGN  12.281384  25000000.0  50000000  Free    0.0   
4           ART_AND_DESIGN   6.874198   2800000.0    100000  Free    0.0   
...                    ...        ...         ...       ...   ...    ...   
10833  BOOKS_AND_REFERENCE   3.784190    619000.0      1000  Free    0.0   
10834               FAMILY   1.945910   2600000.0       500  Free    0.0   
10836               FAMILY   3.637586  53000000.0      5000  Free    0.0   
10837               FAMILY   1.386294   3600000.0       100  Free    0.0   
10840            LIFESTYLE  12.894978  19000000.0  10000000  Free    0.0   

      Content Rating Rating_category  
0           Everyone            High  
1           Everyone            High  
2           Everyone            High  
3               Teen            High  
4           Everyone            High  
...              ...             ...  
10833       Everyone            High  
10834       Everyone            High  
10836       Everyone            High  
10837       Everyone            High  
10840       Everyone            High  

[7424 rows x 8 columns]

Q13) Encode the categorical columns.

#To encode the categorical columns, we are using Label-encoder method as the number of unique values in each feature is a significantly big number and the data would become too big and tough to understand and interpret.

In [99]:
encoded = LabelEncoder()
for col in data1.columns:
    if data1[col].dtype=='object':
        data1[col]=encoded.fit_transform(data1[col])

In [100]:
data1

Category    Reviews        Size  Installs  Type  Price  Content Rating  \
0             0   5.068904  19000000.0     10000     0    0.0               1   
1             0   6.874198  14000000.0    500000     0    0.0               1   
2             0  11.379508   8700000.0   5000000     0    0.0               1   
3             0  12.281384  25000000.0  50000000     0    0.0               4   
4             0   6.874198   2800000.0    100000     0    0.0               1   
...         ...        ...         ...       ...   ...    ...             ...   
10833         3   3.784190    619000.0      1000     0    0.0               1   
10834        11   1.945910   2600000.0       500     0    0.0               1   
10836        11   3.637586  53000000.0      5000     0    0.0               1   
10837        11   1.386294   3600000.0       100     0    0.0               1   
10840        18  12.894978  19000000.0  10000000     0    0.0               1   

       Rating_category  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  
...                ...  
10833                0  
10834                0  
10836                0  
10837                0  
10840                0  

[7424 rows x 8 columns]

Q14) Segregate the target and independent features (Hint: Use Rating_category as the target)

In [105]:
# Segregating the target and independent variables
target = data1['Rating_category']
independent_variables = data1.loc[:, data1.columns != 'Rating_category'] 

In [106]:
target

0        0
1        0
2        0
3        0
4        0
        ..
10833    0
10834    0
10836    0
10837    0
10840    0
Name: Rating_category, Length: 7424, dtype: int32

In [107]:
independent_variables

Category    Reviews        Size  Installs  Type  Price  Content Rating
0             0   5.068904  19000000.0     10000     0    0.0               1
1             0   6.874198  14000000.0    500000     0    0.0               1
2             0  11.379508   8700000.0   5000000     0    0.0               1
3             0  12.281384  25000000.0  50000000     0    0.0               4
4             0   6.874198   2800000.0    100000     0    0.0               1
...         ...        ...         ...       ...   ...    ...             ...
10833         3   3.784190    619000.0      1000     0    0.0               1
10834        11   1.945910   2600000.0       500     0    0.0               1
10836        11   3.637586  53000000.0      5000     0    0.0               1
10837        11   1.386294   3600000.0       100     0    0.0               1
10840        18  12.894978  19000000.0  10000000     0    0.0               1

[7424 rows x 7 columns]

Q15) Split the dataset into test and train

In [110]:
X =  data1.loc[:, data1.columns != 'Rating_category']
Y = data1['Rating_category']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=7)

In [111]:
X_train

Category    Reviews        Size    Installs  Type  Price  Content Rating
6505        17   4.356709    812000.0       10000     0    0.0               1
2722        26   7.413970  33000000.0       50000     0    0.0               1
1700        14  17.137780  76000000.0  1000000000     0    0.0               2
1127        12   8.376551   4500000.0      500000     0    0.0               1
4025        11   5.723585   3300000.0        5000     0    0.0               1
...        ...        ...         ...         ...   ...    ...             ...
8438        11  11.478934  26000000.0     1000000     0    0.0               2
4092        25   8.446771   5900000.0     1000000     0    0.0               1
874          9  14.418888  52000000.0   100000000     0    0.0               1
1931        14  12.902776  70000000.0    10000000     0    0.0               1
6381        18   2.197225   3100000.0        1000     0    0.0               1

[5939 rows x 7 columns]

In [112]:
X_test

Category    Reviews        Size  Installs  Type  Price  Content Rating
5406         29   3.465736  33000000.0     10000     0    0.0               1
5037         24  12.015057  15000000.0  10000000     0    0.0               1
10601        30   4.521789  20000000.0     10000     0    0.0               1
7124          6   6.820016  12000000.0    100000     0    0.0               3
1162         12  11.522846  33000000.0   5000000     0    0.0               1
...         ...        ...         ...       ...   ...    ...             ...
4513         29   8.357963  14000000.0    500000     0    0.0               1
4428         25   6.601230   6800000.0     50000     0    0.0               1
9327          4   3.367296  16000000.0       500     0    0.0               1
41            0   5.170484  10000000.0    100000     0    0.0               1
645           7   0.000000   5500000.0       500     0    0.0               3

[1485 rows x 7 columns]

In [113]:
Y_train

6505    0
2722    0
1700    0
1127    0
4025    0
       ..
8438    0
4092    0
874     0
1931    0
6381    0
Name: Rating_category, Length: 5939, dtype: int32

In [114]:
Y_test

5406     0
5037     0
10601    0
7124     0
1162     0
        ..
4513     0
4428     0
9327     1
41       0
645      1
Name: Rating_category, Length: 1485, dtype: int32

Q16) Standardize the data, so that the values are within a particular range. 

In [116]:
scaler = StandardScaler()

scaled_data = scaler.fit_transform(data1)

In [117]:
scaled_data

array([[-2.03766618, -0.67995305, -0.15992777, ..., -0.06305842,
        -0.46322046, -0.35358018],
       [-2.03766618, -0.19653147, -0.37330014, ..., -0.06305842,
        -0.46322046, -0.35358018],
       [-2.03766618,  1.0099001 , -0.59947486, ..., -0.06305842,
        -0.46322046, -0.35358018],
       ...,
       [-0.68621673, -1.0632313 ,  1.29100439, ..., -0.06305842,
        -0.46322046, -0.35358018],
       [-0.68621673, -1.66608206, -0.81711468, ..., -0.06305842,
        -0.46322046, -0.35358018],
       [ 0.17379656,  1.41571255, -0.15992777, ..., -0.06305842,
        -0.46322046, -0.35358018]])